tests/SelfTest/UsageTests/Matchers.tests.cpp

```diff
@@ -556,65 +556,218 @@ namespace { namespace MatchersTests {
+        >::value);
+
+        REQUIRE_THAT(1, MatcherA() || MatcherB() || MatcherC() || MatcherD());
+    }
+
+    TEST_CASE("Combining MatchAllOfGeneric does not nest", "[matchers][templated]") {
+        STATIC_REQUIRE(std::is_same<
+            decltype(MatcherA() && MatcherB() && MatcherC()),
+            Catch::Matchers::Impl::MatchAllOfGeneric<MatcherA, MatcherB, MatcherC>
+        >::value);
+
+        REQUIRE_THAT(1, MatcherA() && MatcherB() && MatcherC());
 
-    std::array<int, 3> a{ 5, 3, 1 };
+        STATIC_REQUIRE(std::is_same<
+            decltype(MatcherA() && MatcherB() && MatcherC() && MatcherD()),
+            Catch::Matchers::Impl::MatchAllOfGeneric<MatcherA, MatcherB, MatcherC, MatcherD>
+        >::value);
 
-    REQUIRE_THAT(vec,
-                 Predicate<std::vector<int>>([](auto const& vec) {
-                     return std::all_of(vec.begin(), vec.end(), [](int elem) {
-                         return elem % 2 == 1;
-                     });
-                 }, "All elements are odd") &&
-                 !EqualsRange(a));
-}
+        REQUIRE_THAT(1, MatcherA() && MatcherB() && MatcherC() && MatcherD());
+    }
+
+    TEST_CASE("Combining MatchNotOfGeneric does not nest", "[matchers][templated]") {
+        STATIC_REQUIRE(std::is_same<
+            decltype(!MatcherA()),
+            Catch::Matchers::Impl::MatchNotOfGeneric<MatcherA>
+        >::value);
+
+        REQUIRE_THAT(0, !MatcherA());
+
+        STATIC_REQUIRE(std::is_same<
+            decltype(!!MatcherA()),
+            MatcherA const&
+        >::value);
 
-TEST_CASE("Combining concrete matchers does not use templated matchers", "[matchers][templated]") {
-    using Catch::Matchers::StartsWith;
-    using Catch::Matchers::EndsWith;
+        REQUIRE_THAT(1, !!MatcherA());
 
-    STATIC_REQUIRE(std::is_same<decltype(StartsWith("foo") || StartsWith("bar") && EndsWith("bar") && !EndsWith("foo")), Catch::Matchers::Impl::MatchAnyOf<std::string>>::value);
-}
+        STATIC_REQUIRE(std::is_same<
+            decltype(!!!MatcherA()),
+            Catch::Matchers::Impl::MatchNotOfGeneric<MatcherA>
+        >::value);
+
+        REQUIRE_THAT(0, !!!MatcherA());
+
+        STATIC_REQUIRE(std::is_same<
+            decltype(!!!!MatcherA()),
+            MatcherA const &
+        >::value);
+
+        REQUIRE_THAT(1, !!!!MatcherA());
+    }
+
+    struct EvilAddressOfOperatorUsed : std::exception {
+        EvilAddressOfOperatorUsed() {}
+        const char* what() const noexcept override {
+            return "overloaded address-of operator of matcher was used instead of std::addressof";
+        }
+    };
+
+    struct EvilMatcher : Catch::MatcherGenericBase {
+        std::string describe() const override {
+            return "equals: 45";
+        }
+
+        bool match(int i) const {
+            return i == 45;
+        }
+
+        EvilMatcher const* operator& () const {
+            throw EvilAddressOfOperatorUsed();
+        }
+    };
```

Comment:
How about `operator,`? 😈

Reply:
I don't think anything here is vulnerable to that *right now*, but it can't hurt :)